src/proto/prototar/main.go

```diff
@@ -0,0 +1,351 @@
+package main
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/spf13/cobra"
+	"github.com/zeebo/xxh3"
+	"go.uber.org/zap"
+)
+
+// nameAndTarWriter records the filenames of files written to the archive.
+type nameAndTarWriter struct {
+	tar.Writer
+	prefix string
+	names  map[string]struct{}
+}
+
+func (w *nameAndTarWriter) WriteHeader(h *tar.Header) error {
+	if err := w.Writer.WriteHeader(h); err != nil {
+		return errors.Wrap(err, "write underlying header")
+	}
+	if w.names == nil {
+		w.names = make(map[string]struct{})
+	}
+	w.names[path.Join(w.prefix, h.Name)] = struct{}{}
+	return nil
+}
+
+func addToArchive(ctx context.Context, tw *nameAndTarWriter, f fs.FS) error {
+	// TODO(jrockway): For this to be hermetic, WalkDir needs to always visit files in the same
```

Comment:
Simple enough to build up the files in memory and sort them, right?

Reply:
Done.